### PR TITLE
Tabbed components disable routing

### PR DESCRIPTION
**Pull request by djhi**

### CODE REVIEW

**Comment by fzaninotto** [comment]
I merged `master` on `next`, so you'll need to rebase your PR.

File: docs/CreateEdit.md [hunk 1]
```diff
@@ -870,6 +870,7 @@ Here are all the props accepted by the `<TabbedForm>` component:
 * `saving`: A boolean indicating whether a save operation is ongoing. This is passed automatically by `react-admin` when the form component is used inside `Create` and `Edit` components.
 * [`warnWhenUnsavedChanges`](#warning-about-unsaved-changes)
 * [`sanitizeEmptyValues`](#setting-empty-values-to-null)
+* [`syncWithLocation`](#syncwithlocation)
```
Comment (fzaninotto) [inline docs/CreateEdit.md:873]: I don't think the anchor matches the autogenerated section id

File: docs/CreateEdit.md [hunk 2]
```diff
@@ -927,6 +928,45 @@ To style the tabs, the `<FormTab>` component accepts two props:
 - `className` is passed to the tab *header*
 - `contentClassName` is passed to the tab *content*
 
+### Sync With Location
+
+You can also opt out the location synchronization by passing `false` to the `syncWithLocation` prop of the `<TabbedForm>` component, in which case, the `path` prop on the `FormTab` components will be ignored.
```
Comment (fzaninotto) [inline docs/CreateEdit.md:933]:
```suggestion
You can also opt out the location synchronization by passing `false` to the `syncWithLocation` prop of the `<TabbedForm>` component. This allows e.g. to have several `<TabbedForm>` components in a page.
```

File: docs/CreateEdit.md [hunk 3]
```diff
@@ -927,6 +928,45 @@ To style the tabs, the `<FormTab>` component accepts two props:
...
+                        <EditButton />
+                    </Datagrid>
+                </ReferenceManyField>
+            </FormTab>
+        </TabbedForm>
+    </Edit>
+);
+```
+{% endraw %}
+
```
Comment (fzaninotto) [inline docs/CreateEdit.md:969]:
```suggestion
**Tip**: When `syncWithLocation` is `false`, the `path` prop of the `<FormTab>` components is ignored.
```

File: docs/Show.md [hunk 4]
```diff
@@ -332,6 +332,45 @@ To style the tabs, the `<Tab>` component accepts two props:
 - `className` is passed to the tab *header*
 - `contentClassName` is passed to the tab *content*
 
+You can also opt out the location synchronization by passing `false` to the `syncWithLocation` prop of the `<TabbedShowLayout>` component, in which case, the `path` prop on the `Tab` components will be ignored.
```
Comment (fzaninotto) [inline docs/Show.md:335]:
```suggestion
You can also opt out the location synchronization by passing `false` to the `syncWithLocation` prop of the `<TabbedShowLayout>` component. This allows e.g. to have several `<TabbedShowLayout>` components in a page.
```

File: docs/Show.md [hunk 5]
```diff
@@ -332,6 +332,45 @@ To style the tabs, the `<Tab>` component accepts two props:
...
+                        <EditButton />
+                    </Datagrid>
+                </ReferenceManyField>
+            </Tab>
+        </TabbedShowLayout>
+    </Show>
+);
+```
+{% endraw %}
+
```
Comment (fzaninotto) [inline docs/Show.md:373]:
```suggestion
**Tip**: When `syncWithLocation` is `false`, the `path` prop of the `<Tab>` components is ignored.
```

File: packages/ra-ui-materialui/src/detail/Tab.tsx [hunk 6]
```diff
@@ -126,7 +129,7 @@ Tab.propTypes = {
     icon: PropTypes.element,
     label: PropTypes.string.isRequired,
     path: PropTypes.string,
-    value: PropTypes.string,
+    value: PropTypes.any,
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/Tab.tsx:132]: Why did you change that?

File: packages/ra-ui-materialui/src/detail/Tab.tsx [hunk 7]
```diff
@@ -73,7 +74,9 @@ const Tab = ({
             value={value}
             icon={icon}
             className={classnames('show-tab', className)}
-            {...({ component: Link, to: value } as any)} // to avoid TypeScript screams, see https://github.com/mui-org/material-ui/issues/9106#issuecomment-451270521
+            {...(syncWithLocation
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/Tab.tsx:77]: I find it weird that you handle the tab click here when `syncWithLocation` is true, and in the parent component when it's false.

I think you'd better move the tab click handling in the parent component altogether.

Same in formTab.

File: packages/ra-ui-materialui/src/detail/TabbedShowLayoutTabs.tsx [hunk 8]
```diff
@@ -1,24 +1,34 @@
...
-import Tabs from '@material-ui/core/Tabs';
+import Tabs, { TabsProps } from '@material-ui/core/Tabs';
 import { useLocation, useRouteMatch } from 'react-router-dom';
+import { TabProps } from './Tab';
 
 export const getTabFullPath = (tab, index, baseUrl) =>
     `${baseUrl}${
         tab.props.path ? `/${tab.props.path}` : index > 0 ? `/${index}` : ''
-    }`;
+    }`.replace('//', '/');
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/TabbedShowLayoutTabs.tsx:11]: what's that for? a comment would help

File: packages/ra-ui-materialui/src/form/FormTabHeader.tsx [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+            value={value}
+            icon={icon}
+            className={classnames('form-tab', className, {
+                [classes.errorTabButton]:
+                    formGroup.invalid &&
+                    formGroup.touched &&
+                    location.pathname !== value,
+            })}
+            component={syncWithLocation ? Link : undefined}
+            to={{ ...location, pathname: value }}
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/form/FormTabHeader.tsx:36]: does the `to` prop make sense if `syncWithLocation` is false?

File: packages/ra-ui-materialui/src/form/TabbedForm.spec.tsx [hunk 10]
```diff
@@ -8,10 +9,10 @@ import {
 } from 'ra-core';
 import { renderWithRedux } from 'ra-test';
 
-import TabbedForm from './TabbedForm';
-import FormTab from './FormTab';
+import { TabbedForm } from './TabbedForm';
+import { FormTab } from './FormTab';
 import TextInput from '../input/TextInput';
-import { fireEvent } from '@testing-library/react';
+import { fireEvent, isInaccessible, waitFor } from '@testing-library/react';
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/form/TabbedForm.spec.tsx:15]: Linter warning

File: packages/ra-ui-materialui/src/form/TabbedFormTabs.tsx [hunk 11]
```diff
@@ -69,12 +76,13 @@ export const getTabFullPath = (
 ): string =>
     `${baseUrl}${
         tab.props.path ? `/${tab.props.path}` : index > 0 ? `/${index}` : ''
-    }`;
+    }`.replace('//', '/');
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/form/TabbedFormTabs.tsx:79]: same, please explain

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: packages/ra-ui-materialui/src/detail/Tab.tsx (hunk 6 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/detail/Tab.tsx:132]: Because we can use path or indexes which are numbers. I should probably be more specific though. However, MUI value for the `Tabs` is a TS any

File: packages/ra-ui-materialui/src/detail/Tab.tsx (hunk 7 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/detail/Tab.tsx:77]: That's because it's a simple link when `syncWithLocation` is `true`. I'd rather avoid having buttons calling the router methods as it is less accessible imo

**Review by fzaninotto — APPROVED** [review]

**Comment by fzaninotto** [comment]
tests fail

**Review by fzaninotto — APPROVED** [review]
